test/functional/zmq_test.py

```diff
@@ -13,9 +13,33 @@
                                  hash256,
                                 )
 
+class ZMQSubscriber:
+    def __init__(self, socket, address):
+        self.address = address
+        self.socket = socket
+        self.sequence = 0
+
 class ZMQTest (BitcoinTestFramework):
     def set_test_params(self):
+        self.address = "tcp://127.0.0.1:28332"
+        self.num_blocks = 5
         self.num_nodes = 2
+        self.subscribers = {}
+
+    def subscribe(self, type):
+        import zmq
+        self.socket.setsockopt(zmq.SUBSCRIBE, type.encode('latin-1'))
+        self.subscribers[type] = ZMQSubscriber(self.socket, self.address)
+
+    def receive(self, type):
```

Comment:
nit: assert_receive? My initial thought was that you're selectively receiving a type (which makes sense if different sockets, but if the sockets are all identical, you get what you're given, as this code tests)

Reply:
Below feels better to read `foo = self.receive(...)`, no need for the `assert_` IMO.

> but if the sockets are all identical, you get what you're given

Not really, if below you swap 2 consecutive receives the test will fail. What comes from the socket has a specific order and we must follow it when reading.